src/event.rs

```diff
@@ -97,210 +94,218 @@ pub enum Key {
                 }
                 Some(Ok(b'[')) => {
                     // This is a CSI sequence.
-                    match iter.next() {
-                        Some(Ok(b'D')) => Event::Key(Key::Left),
-                        Some(Ok(b'C')) => Event::Key(Key::Right),
-                        Some(Ok(b'A')) => Event::Key(Key::Up),
-                        Some(Ok(b'B')) => Event::Key(Key::Down),
-                        Some(Ok(b'H')) => Event::Key(Key::Home),
-                        Some(Ok(b'F')) => Event::Key(Key::End),
-                        Some(Ok(b'M')) => {
-                            // X10 emulation mouse encoding: ESC [ CB Cx Cy (6 characters only).
-                            let cb = iter.next().unwrap().unwrap() as i8 - 32;
-                            // (1, 1) are the coords for upper left.
-                            let cx = (iter.next().unwrap().unwrap() as u8).saturating_sub(32) as u16;
-                            let cy = (iter.next().unwrap().unwrap() as u8).saturating_sub(32) as u16;
-                            Event::Mouse(match cb & 0b11 {
-                                0 => {
-                                    if cb & 0x40 != 0 {
-                                        MouseEvent::Press(MouseButton::WheelUp, cx, cy)
-                                    } else {
-                                        MouseEvent::Press(MouseButton::Left, cx, cy)
-                                    }
-                                }
-                                1 => {
-                                    if cb & 0x40 != 0 {
-                                        MouseEvent::Press(MouseButton::WheelDown, cx, cy)
-                                    } else {
-                                        MouseEvent::Press(MouseButton::Middle, cx, cy)
-                                    }
-                                }
-                                2 => MouseEvent::Press(MouseButton::Right, cx, cy),
-                                3 => MouseEvent::Release(cx, cy),
-                                _ => return error,
-                            })
-                        },
-                        Some(Ok(b'<')) => {
-                            // xterm mouse encoding:
-                            // ESC [ < Cb ; Cx ; Cy ; (M or m)
-                            let mut buf = Vec::new();
-                            let mut c = iter.next().unwrap().unwrap();
-                                while match c {
-                                    b'm' | b'M' => false,
-                                    _ => true,
-                                } {
-                                    buf.push(c);
-                                    c = iter.next().unwrap().unwrap();
-                                }
-                            let str_buf = String::from_utf8(buf).unwrap();
-                            let nums = &mut str_buf.split(';');
+                    parse_csi(iter).ok_or(error)?
```

Comment:
This uses the new stealth try! alternative.
We could replace it with a conventiona `try!` if prefered.